### DISCUSSION TITLE
How to access to the html content during the build process?

### DISCUSSION THREAD

**Discussion by ivodolenc** [sveltejs/kit#7324] [General]
Hi,

Is it possible to inject data into the `<head></head>` section during the build process, but from the `vite-plugin`?

Specifically, I'm building a custom `vite-plugin` for SvelteKit static projects.

Plugin will start only during build process (so when user want to generate static project for production, it would run `npm run build` and plugin will start only then). I will manage this by vite `configResolve` with simple conditionally check inside of hook, nothing special.

I just need to know how to access to SvelteKit html content during build so the plugin can automatically inject `<link>` tag to `<head>` section for all pages.

If it will help, this is a `manifest.json` that will be automatically generated with all dynamic content during the `build` process as it is not needed in dev mode.

The tricky part is that the manifest need to be linked inside `<head>` and I just need to inject that link through plugin.

I want to simplify the use, so the user will install the plugin and activate it and that's it. Plug & Play.

### Additional info:

- Plugin is designed for SSG mode (adapter-static)
- Injection is only required during build mode (have solution for that)
- The tag should be imported in all pages
- Also, I want to do it from plugin to simplify the user workflow (this can be done on user side by adding link tag in `layout.svelte` but then would need some conditionally check if is the `dev` or `build` mode, it can be messy or confusing for the user)

### Vite solution

I think that for Vite's vanilla projects, there is `transformIndexHtml` hook. So this would replace `</head>` end tag with new `<link /></head>` tags:

```js
// Example from Vite's docs

const htmlPlugin = () => {
  return {
    name: 'html-transform',

    transformIndexHtml(html) {
      // replace '</head>' with '<link /></head>'
      return html.replace(/<\/head>/, `<link id="..." href="...." /></head>`)
    }
  }
}
```

Is something like this posible in SvelteKit but from `vite-plugin`?

Maybe some Vite or RollUp hook? I think that someone is already ask something similar on Github and Rich is reply that SvelteKit uses internal building system for static rendering in adapter-static.

Any advice?

**Comment by MentalGear** [comment]
I think you should be able to do this with vite only by adding your plugin after sveltekit, use vite's transformIndexHTML - also for plugins there's a flag to only run on build.

**Comment by dominikg** [comment]
sveltekit does not support transformIndexHtml hook of vite as it uses it's own implementation for building the html.

To manipulate the final html during SSR, you can use transformPageChunk of resolve, see https://kit.svelte.dev/docs/accessibility#the-lang-attribute for an example.

In case of head, you can also explore using svelte:head in appropriate layouts.

**Comment by ivodolenc** [comment]
Thanks for the info.

> sveltekit does not support transformIndexHtml hook of vite as it uses it's own implementation for building the html.

Yeah that is unfortunately bad news. I hope that in the future SvelteKit will provide an API to support external solutions such as `vite-plugins` or even Svelte own `plugin system`.

This is essential for plugin authors/contributors because if we can't manipulate with the content outside of user's workspace, such plugins are practically useless (of course, not entirely).

The main goal of external plugins is to simplify user workflow and share a great development experience across multiple projects. **Plug & Play** 👍

For example, it's easier to create plugin and publish it on the npm and later import it in every project. Others can contribute and help with usage.

It's not practical to have hooks.js files inside every projects just for common tasks. Also, maintaining it could become quite difficult.

> To manipulate the final html during SSR, you can use transformPageChunk of resolve, see https://kit.svelte.dev/docs/accessibility#the-lang-attribute for an example.

Can we use `transformPageChunk` through `vite-plugin`?

### Alternative solution

So far I've managed to solve it with a workaround. Not ideal, but works ok. Of course, this is a simplified example just for demonstration.

```js
const Plugin = () => {
  return {
    name: 'vite-plugin',

    apply(config, { command }) {
      // apply only on build but not for SSR
      return command === 'build' && !config.build.ssr
    },

    /**
     * Opens the app.html file and injects the link into the <head>
     */
    buildStart() {
      let appTemplate = 'src/app.html'
      let link = `<link /></head>`
      let appContent = readFileSync(appTemplate, 'utf-8')
      let injectLink = appContent.replace(/<\/head>/, link)

      writeFileSync(appTemplate, injectLink)
    },

    /**
     * When building is done, opens the app.html file
     * and removes the link from the <head>
     */
    closeBundle() {
      let appTemplate = 'src/app.html'
      let link = `<link />`
      let appContent = readFileSync(appTemplate, 'utf-8')
      let removeLink = appContent.replace(link, '')

      writeFileSync(appTemplate, removeLink)
    }
  }
}
```
Comments and advice are welcome.
